### ISSUE TITLE
Enabling Global Optimizer requires Test Run on cross-build distributions

### ISSUE THREAD

**Issue by liontiger23** [excelsior-oss/excelsior-jet-maven-plugin#81]
The note at the end of [Global Optimizer](https://github.com/excelsior-oss/excelsior-jet-maven-plugin/wiki/Optimization-Settings#global-optimizer) section of the wiki says:

> Note: performing a Test Run is mandatory if the Global Optimizer is enabled.

And the plugin refuses to build with global without Test Run, even on cross-build distributions (e.g. Linux/ARM), where Test Run is currently not supported, but Global Optimizer is.

Naturally, plugins should support this use case (using Global Optimizer without Test Run on Linux/ARM) as well.

**Comment by pjBooms** [comment]
It was done intentionally. Without Test Run, Global Optimizer may worsen startup and performance. It is enough to perform Test Run on one platform and use the results on another (as reflective usage is mostly platform neutral)

**Comment by lbeuster** [comment]
I think the 2 JET installations are only a workaround. The root problem seems to me that the JET embedded installation is not complete. 

If I understood everything right then a test run is required for full functionality (in this case optimization). But a test run is not possible, so the installation is incomplete.
Having another JET installation for the host system is only a workaround because in that case I may have different versions of JET in one build (one licensed version with hotfixes and one eval version). This again may lead to overhead on your side because you have to ensure that these different versions can work together. And needing an eval version although I have a valid license is strange ;)

But if the 2 installations are only a workaround then I would not support that workaround in the maven plugin. It should be possible now to define 2 different plugin execution configurations in the pom (one for build and one for testrun) with different JET homes (I could try that next week).

**Comment by pjBooms** [comment]
> In order to perform Test Run with a different version of JET, you will still need to modify your JET configuration just to perform `jet:testrun` and keep a JET for it up-to-date.
> It would be more convenient to have a separate task something like `jet:hosttestrun` and a separate property `hostJetHome` where the location of host-platform JET is stored. This way in order to perform Test Run with a different version of JET you won't need to change any of JET configurations.

You don't need to modify the configuration. You just need to switch jet home (that is not necessary to be present in the configuration). Read 
https://github.com/excelsior-oss/excelsior-jet-maven-plugin/wiki/Prerequisites#excelsior-jet-installation-directory-lookup for details. Your case is supported right now with

`jet test:run -Djet.home=<HOST_JET>`

**Comment by lbeuster** [comment]
With config in the pom you can build everything with one maven invocation (for CI/Jenkins). In that case you can call "mvn jet:testrun jet:build".

**Comment by pjBooms** [comment]
I would comment the issue later. It is not easy as can be thought :)

**Comment by pjBooms** [comment]
I think that `jet:testrun` task should create such an image for cross compilation JET with instructions how to perform Test Run on the target and where to copy resulting execution profile.
BTW, `jet:profile` task follows exactly this way: it creates profiling image that you need to just copy and run on the target to get the execution profile that you copy to the host then enabling PGO.  I think that `jet:testrun` and `jet:profile` should create identical images for gathering all the profiles at once. 

On the other hand, I agree that performing the Test Run on a target is not so convenient when it is actually possible to perform it on the host because it cannot be automated anyhow (you need to do some manual actions for that). However, ability to perform Test Run on the host actually means that cross-compilation JET should have two JREs inside anyway: one for the host and another for the target.
It cannot be solved right now.

So for you, @lbeuster,  I see three possibilities to workaround or resolve the issue:

1. We can make modifications in the Maven plugin for `jet:testrun` to create profiling image for the target.

2. We can provide you a retail version of JET (say JET Embedded for AMD64 with a special license limiting the usage of it just for Test Run).

3. You can also live with JET Evaluation workaround. Actually Test Run results it produces are equal with retail version. 

Please let us know what you think.

**Comment by lbeuster** [comment]
Point (3) with the eval version is completely enough for me as a workaround. Unfortunately jet:testrun fails with a system exception on our project on Linux and OSX. But that's another story... ;)

**Comment by lbeuster** [comment]
Unfortunately I was not able to configure the plugin to be executed with different configs and goals in the same maven invocation. Maybe it's because it forks itself a maven process?
In the end only the last execution config was really executed. Is there any trick?

If not, maybe it's possible to change that default behavior and behave like other plugins or add additional targets to the plugin that can be integrated into the normal build lifecycle to make the following code work:

```
<plugin>
    <groupId>com.excelsiorjet</groupId>
    <artifactId>excelsior-jet-maven-plugin</artifactId>
    <executions>

        <execution>
            <id>testrun</id>
            <goals><goal>testrun</goal></goals>
            <phase>package</phase>
            <configuration>
                <jetHome>/path/to/jet-linux</jetHome>
            </configuration>
        </execution>

        <execution>
            <id>build</id>
            <goals><goal>build</goal></goals>
            <phase>package</phase>
            <configuration>
                <jetHome>/path/to/jet-embedded</jetHome>
            </configuration>
        </execution>

    </executions>
</plugin>
```

**Comment by pjBooms** [comment]
>  Maybe it's because it forks itself a maven process?
No. There is nothing special with the plugin.

>>In the end only the last execution config was really executed. Is there any trick?
I applied your "executions" to a simple project and I observe another behavior. 
In my environment I see that "build" task is executed first then "testrun" task is executed (with applied jetHome) and then "build" task is executed again. Do you really see only the "build" task execution?
BTW, what Maven version do you use?
Anyway, even what I observe is not expected. We need some time to investigate the issue (probably it is better to create a separate issue for this). As the plugin is open source, you are welcome to make investigations in parallel and share your results. 
From Internet, I see that similar problems (with twice goals executions) happen sometimes for other plugins and they can be provoked by other plugins applied.

Probably, it can be fixed with another "defaultPhase" for the plugin goals (unsure), anyway it is some Maven specific, IMHO.

**Comment by lbeuster** [comment]
I don't think it's a maven issue. The jet plugin does something special because during the jet plugin execution with "mvn jet:build" the plugin spawns it's own maven lifecycle (or however it is called) including the execution of other plugins (resources, compiler, surefire). Maybe there's a reason for doing that, but I think the maven way is to integrated it into the regular maven lifecycle as any other plugin (e.g. compiler). 
Doing a "mvn jet:build" should only execute the jet plugin, not any other plugins. And doing a regular  "mvn verify" should execute all regular lifecycle plugins including the jet plugin - if it has explicitly integrated into the plugin section of the pom.xml.

With the current behavior it's also not possible to do a "mvn jet:build" in the root of a multi-maven-project because it tries to do the jet build in every sub project. This means I have to first do a "mvn install" on the root and then do a mvn "jet:build" on the sub project.

```
$ mvn jet:build
[INFO] -------------------< mytest.jet-exe >-------------------
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] >>> excelsior-jet-maven-plugin:1.3.1:build (default-cli) > package @ jvs-exe >>>
[INFO] --- maven-enforcer-plugin:3.0.0-M2:enforce (enforce-banned-dependencies) @ jvs-exe ---
[INFO] --- git-commit-id-plugin:2.2.6:revision (get-the-git-infos) @ jvs-exe ---
[INFO] --- maven-resources-plugin:3.0.2:resources (default-resources) @ jvs-exe ---
[INFO] --- maven-compiler-plugin:3.8.0:compile (default-compile) @ jvs-exe ---
[INFO] --- maven-resources-plugin:3.0.2:testResources (default-testResources) @ jvs-exe ---
[INFO] --- maven-compiler-plugin:3.8.0:testCompile (default-testCompile) @ jvs-exe ---
[INFO] --- maven-surefire-plugin:2.22.1:test (default-test) @ jvs-exe ---
[INFO] --- maven-jar-plugin:3.1.1:jar (default-jar) @ jvs-exe ---
[INFO] <<< excelsior-jet-maven-plugin:1.3.1:build (default-cli) < package @ jvs-exe <<<
--------------------------------------------
everything above this line should not be done by the plugin
--------------------------------------------
[INFO] --- excelsior-jet-maven-plugin:1.3.1:build (default-cli) @ jvs-exe ---
```

To fix this and not break the old behavior a new target could be created for every existing target that only executes the plugin itself. The new targets can be integrated into the pom the maven way as in the above XML on top of this issue.
Or maybe easier to add a config param to switch the special behavior on/off (if this is possible).

**Comment by pjBooms** [comment]
>>The jet plugin does something special 
If you point me the place where it does it, I would be greatly appreciated.

> With the current behavior it's also not possible to do a "mvn jet:build" in the root of a multi-maven-project because it tries to do the jet build in every sub project.

It is common Maven behavior. Every other plugins behave the same way. E.g., if you have spring boot plugin in the root project, Maven will try to apply it for every sub project.

>>  a new target could be created for every existing target
Do you know some other plugins that do it?

**Comment by lbeuster** [comment]
> > The jet plugin does something special
> 
> If you point me the place where it does it, I would be greatly appreciated.
> I rather think that the plugin does not do something in addition to stop Maven doing the things that it does but unfortunately I do not know what it does not do.

The plugin does too much, can you see this from the above logs? 
For me it looks like the plugin "somehow forces" maven to first execute the "package" lifecycle and after that the regular plugin goal is executed:
```
[INFO] >>> excelsior-jet-maven-plugin:1.3.1:build (default-cli) > package @ jvs-exe >>>
...
[INFO] <<< excelsior-jet-maven-plugin:1.3.1:build (default-cli) < package @ jvs-exe <<<
[INFO] --- excelsior-jet-maven-plugin:1.3.1:build (default-cli) @ jvs-exe ---
```
 
> > With the current behavior it's also not possible to do a "mvn jet:build" in the root of a multi-maven-project because it tries to do the jet build in every sub project.
> 
> It is common Maven behavior. Every other plugins behave the same way. E.g., if you have spring boot plugin in the root project, Maven will try to apply it for every sub project.

Yes, that's true. But it also fits into the normal build cycle - and that's the preferred way to use it. The point is that because of the special things the jet plugin does, it's not possible to build a multi-maven project with one maven invocation (especially if I need 2 jet plugin targets: testrun + build). Furthermore I even have to install the artifact into the local maven repo. so that the jet:build can find them. 

> > a new target could be created for every existing target
> 
> Do you know some other plugins that do it?

No, because it's not nice. It would be only useful to go away from the special things forward to the maven way and keep backward compatibility. But I would prefer the config flag to achieve this. Or do not support the special things anymore ;)

**Comment by pjBooms** [comment]
> everything above this line should not be done by the plugin

Why do you think that it is done by plugin? It is done by Maven. Why it does it, should be investigated of course.

>For me it looks like the plugin "somehow forces

The plugin code is here:

https://github.com/excelsior-oss/excelsior-jet-maven-plugin/blob/master/src/main/java/com/excelsiorjet/maven/plugin/BuildMojo.java

It is really that simple and does not call Maven itself. Only Maven calls it.

**Comment by lbeuster** [comment]
Of course it is done by maven but the plugin tells maven to do it.

Thanks for the link to the code, the first annotation of the plugin is:
```
@Execute(phase = LifecyclePhase.PACKAGE)
```

The maven doc at https://maven.apache.org/plugin-tools/maven-plugin-annotations/apidocs/org/apache/maven/plugins/annotations/Execute.html states:
```
Used if your Mojo needs to fork a lifecycle.
```

That's the reason for the special things.

For instance the CompilerMojo (https://github.com/apache/maven-compiler-plugin/blob/master/src/main/java/org/apache/maven/plugin/compiler/CompilerMojo.java) doesn't have this annotation.

**Comment by pjBooms** [comment]
>That's the reason for the special things.

Great. Please create a separate issue for this. I will test the plugin without the annotation and release a new version if everything is ok.

**Comment by lbeuster** [comment]
Thanks, #82 created.
